### PR TITLE
Overmap and mapgen: testing, cleanup, and documentation

### PR DESCRIPTION
**Pull request by wapcaplet**
#### Summary

```SUMMARY: Infrastructure "Overmap and mapgen: testing, cleanup, and documentation"```

#### Purpose of change

Some of the procedural map generation code (for example `map::draw_lab` in `mapgen.cpp`) is long and complicated, too deeply nested, poorly documented, largely untested, and/or in dire want of refactoring or JSONizing.

With this PR I hope to include more tests, code comments, and documentation, to pave the way for future refactoring. Code refactoring is mostly *outside* the scope of this PR; my intention here is to document current behavior, before meddling in the code in ways that might change that behavior.

#### Describe the solution

**Completely** overhaul `doc/MAPGEN.md`.

Also includes new test cases:

- `connects_to` unit tests in `tests/mapgen_function_test.cpp`
- `is_ot_match` unit tests in `tests/overmap_test.cpp`

#### Describe alternatives you've considered

Refactoring before having tests in place - risky.

#### Testing

`tests/cata_test connects_to,is_ot_match`

#### Additional context

~~WIP with *comments requested*. I am learning about overmaps and mapgen as I go, and have no expertise in this part of the game. Corrections and suggestions are most welcome.~~

~~This PR is somewhat open-ended, without a clear definition of "done", but at minimum I want to open the door for future contributors to understand the procedural mapgen code better, and have some tests to run when they (or I) do any refactoring.~~

Reformatting the `MAPGEN.md` alone made me "done" with this PR; I had ambitions for more testing and code comments, but those can come in follow-up work.

### CODE REVIEW

**Comment by damien** [comment]
This is a _substantial_ improvement on he mapgen documentation. Thank you! 😍

File: doc/MAPGEN.md [hunk 1]
```diff
@@ -233,241 +249,307 @@ Example:
...
-Example: [ 1, 3 ] - apply 1-3 times
 
 ### 2.2.2 "square"
 **required** Define a square of things. Requires "x", "y", "x2", "y2", "id" (or "amount" for "radiation")
-> Value: "terrain", "furniture", "trap", "radiation"
+Value: `"terrain", "furniture", "trap", "radiation"`
+
+Example:
+```json
+{ "square": "radiation", "amount": 10, "x:" [ 0, 5 ], "y": [ 0, 5 ], "x2": [ 18, 23 ], "y2": [ 18, 23 ] }
```
Comment (AMurkin) [inline doc/MAPGEN.md]:
```suggestion
{ "square": "radiation", "amount": 10, "x": [ 0, 5 ], "y": [ 0, 5 ], "x2": [ 18, 23 ], "y2": [ 18, 23 ] }
```

Comment (wapcaplet) [reply doc/MAPGEN.md]: Thanks, yeah I noticed a few of those (in fact my vim markdown plugin highlighted them for me but I didn't see why at first). Resolved in the latest push.

**Comment by wapcaplet** [comment]
All I wanted to do was remove the section numbering from `MAPGEN.md` but it turned into a complete rework of the entire document.

image

My main objective was to improve readability of both the raw Markdown and prettified HTML formats, without changing the meaning of anything. I did a lot of minor rephrasing and major reformatting.

Please submit corrections if I have lost anything in the translation.

**Comment by damien** [comment]
Off topic: It may be worth drawing up a short set of style and formatting guidelines for C:DDA markdown docs to keep things consistent going forward. This particular PR is great and I'd love to see the style here used as a baseline for other documents.

**Comment by AMurkin** [comment]
I like tables markup in doc/MAPGEN.md.
In doc/OVERMAP.md each row have trailing spaces and a pipe.
Perhaps it’s worth removing them so that when editing one row, you do not have to adjust each row.

**Pull request merged**